Storage/src/SigningHelper.php

```diff
@@ -0,0 +1,578 @@
+     * @return int
+     * @throws \InvalidArgumentException If an invalid value is given.
+     */
+    protected function normalizeExpiration($expires)
+    {
+        if ($expires instanceof Timestamp) {
+            $seconds = $expires->get()->format('U');
+        } elseif ($expires instanceof \DateTimeInterface) {
+            $seconds = $expires->format('U');
+        } elseif (is_numeric($expires)) {
+            $seconds = (int) $expires;
+        } else {
+            throw new \InvalidArgumentException('Invalid expiration.');
+        }
+
+        return $seconds;
+    }
+
+    /**
+     * Normalizes and encodes the resource identifier.
+     *
+     * @param string $resource The resource identifier. In form
+     *        `[/]$bucket/$object]`.
+     * @return string The resource, with pieces encoded and prefixed with a
+     *        forward slash.
+     */
+    protected function normalizeResource($resource)
+    {
+        $pieces = explode('/', trim($resource, '/'));
+        array_walk($pieces, function (&$piece) {
+            $piece = rawurlencode($piece);
+        });
+        return '/' . implode('/', $pieces);
+    }
+
+    /**
+     * Fixes the user input options, filters and validates data.
+     *
+     * @param array $options Signed URL configuration options.
+     * @return array
+     * @throws \InvalidArgumentException
+     */
+    protected function normalizeOptions(array $options)
+    {
+        $options += [
+            'method' => 'GET',
+            'cname' => self::DEFAULT_DOWNLOAD_HOST,
+            'contentMd5' => null,
+            'contentType' => null,
+            'headers' => [],
+            'saveAsName' => null,
+            'responseDisposition' => null,
+            'responseType' => null,
+            'keyFile' => null,
+            'keyFilePath' => null,
+            'allowPost' => false,
+            'forceOpenssl' => false,
+            'queryParams' => [],
+            'timestamp' => null
+        ];
+
+        $allowedMethods = ['GET', 'PUT', 'POST', 'DELETE'];
+        $options['method'] = strtoupper($options['method']);
+        if (!in_array($options['method'], $allowedMethods)) {
+            throw new \InvalidArgumentException('$options.method must be one of `GET`, `PUT` or `DELETE`.');
+        }
+
+        if ($options['method'] === 'POST' && !$options['allowPost']) {
+            throw new \InvalidArgumentException(
+                'Invalid method. To create an upload URI, use StorageObject::signedUploadUrl().'
+            );
+        }
+        unset($options['allowPost']);
+
+        // For backwards compatibility, strip protocol from cname.
+        $cnameParts = explode('//', $options['cname']);
+        if (count($cnameParts) > 1) {
+            $options['cname'] = $cnameParts[1];
+        }
+
+        $options['cname'] = trim($options['cname'], '/');
+
+        // If a timestamp is provided, use it in place of `now` for v4 URLs only..
+        if ($options['timestamp']) {
```

Comment:
I might have missed it, but I had trouble finding anywhere this is documented.

Reply:
It's undocumented purposely. One should only create a signed URL from the current moment, but for the conformance tests, it is necessary to start from a given time in order to predict the correct signature string. This option is used for testing, since afaik, PHP doesn't let you mock `time()`.

Reply:
Ah, okay. I wasn't sure given the comment in the exception which states "User-provided". Having this for testing is fine.

Reply:
I added to the comment to make clear what the purpose of this is.